# Import libraries

In [34]:
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt 
import seaborn as sns 
%matplotlib inline

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder,LabelEncoder
from sklearn.preprocessing import RobustScaler
from sklearn.naive_bayes import GaussianNB
from sklearn import set_config
from sklearn.metrics import accuracy_score,classification_report


In [25]:
pd.set_option('display.max_columns',100)
set_config(print_changed_only = False)

In [3]:
df = pd.read_csv("bank-additional-full.csv",sep = ';')
df.head()

age        job  marital    education  default housing loan    contact  \
0   56  housemaid  married     basic.4y       no      no   no  telephone   
1   57   services  married  high.school  unknown      no   no  telephone   
2   37   services  married  high.school       no     yes   no  telephone   
3   40     admin.  married     basic.6y       no      no   no  telephone   
4   56   services  married  high.school       no      no  yes  telephone   

  month day_of_week  duration  campaign  pdays  previous     poutcome  \
0   may         mon       261         1    999         0  nonexistent   
1   may         mon       149         1    999         0  nonexistent   
2   may         mon       226         1    999         0  nonexistent   
3   may         mon       151         1    999         0  nonexistent   
4   may         mon       307         1    999         0  nonexistent   

   emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0           1.1          93.994          -36.4      4.857       5191.0  no  
1           1.1          93.994          -36.4      4.857       5191.0  no  
2           1.1          93.994          -36.4      4.857       5191.0  no  
3           1.1          93.994          -36.4      4.857       5191.0  no  
4           1.1          93.994          -36.4      4.857       5191.0  no

1 - age (numeric)

2 - job : type of job (categorical: "admin.","blue-collar","entrepreneur","housemaid","management","retired","self-employed","services","student","technician","unemployed","unknown")

3 - marital : marital status (categorical: "divorced","married","single","unknown"; note: "divorced" means divorced or widowed)

4 - education (categorical: "basic.4y" ,"basic.6y" ,"basic.9y" ,"high.school" ,"illiterate" ,"professional.course"      ,"university.degree" ,"unknown")

5 - default: has credit in default? (categorical: "no","yes","unknown")
   
6 - housing: has housing loan? (categorical: "no","yes","unknown")

7 - loan: has personal loan? (categorical: "no","yes","unknown")
   
# related with the last contact of the current campaign:
   
8 - contact: contact communication type (categorical: "cellular","telephone") 
   
9 - month: last contact month of year (categorical: "jan", "feb", "mar", ..., "nov", "dec")
  
10 - day_of_week: last contact day of the week (categorical: "mon","tue","wed","thu","fri")
  
11 - duration: last contact duration, in seconds (numeric). Important note:  this attribute highly affects the output target (e.g., if duration=0 then y="no"). Yet, the duration is not known before a call is performed. Also, after the end of the call y is obviously known. Thus, this input should only be included for benchmark purposes and should be discarded if the intention is to have a realistic predictive model.
   
# other attributes:
  
12 - campaign: number of contacts performed during this campaign and for this client (numeric, includes last contact)
  
13 - pdays: number of days that passed by after the client was last contacted from a previous campaign (numeric; 999 means client was not previously contacted)
  
14 - previous: number of contacts performed before this campaign and for this client (numeric)
  
15 - poutcome: outcome of the previous marketing campaign (categorical: "failure","nonexistent","success")
   
# social and economic context attributes
  
16 - emp.var.rate: employment variation rate - quarterly indicator (numeric)
  
17 - cons.price.idx: consumer price index - monthly indicator (numeric)     
  
18 - cons.conf.idx: consumer confidence index - monthly indicator (numeric)     
  
19 - euribor3m: euribor 3 month rate - daily indicator (numeric)
  
20 - nr.employed: number of employees - quarterly indicator (numeric)

Output variable (desired target):
  
21 - y - has the client subscribed a term deposit? (binary: "yes","no")

In [4]:
df.shape

(41188, 21)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  object 
 2   marital         41188 non-null  object 
 3   education       41188 non-null  object 
 4   default         41188 non-null  object 
 5   housing         41188 non-null  object 
 6   loan            41188 non-null  object 
 7   contact         41188 non-null  object 
 8   month           41188 non-null  object 
 9   day_of_week     41188 non-null  object 
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  object 
 15  emp.var.rate    41188 non-null  float64
 16  cons.price.idx  41188 non-null  float64
 17  cons.conf.idx   41188 non-null 

In [6]:
df.describe()

age      duration      campaign         pdays      previous  \
count  41188.00000  41188.000000  41188.000000  41188.000000  41188.000000   
mean      40.02406    258.285010      2.567593    962.475454      0.172963   
std       10.42125    259.279249      2.770014    186.910907      0.494901   
min       17.00000      0.000000      1.000000      0.000000      0.000000   
25%       32.00000    102.000000      1.000000    999.000000      0.000000   
50%       38.00000    180.000000      2.000000    999.000000      0.000000   
75%       47.00000    319.000000      3.000000    999.000000      0.000000   
max       98.00000   4918.000000     56.000000    999.000000      7.000000   

       emp.var.rate  cons.price.idx  cons.conf.idx     euribor3m   nr.employed  
count  41188.000000    41188.000000   41188.000000  41188.000000  41188.000000  
mean       0.081886       93.575664     -40.502600      3.621291   5167.035911  
std        1.570960        0.578840       4.628198      1.734447     72.251528  
min       -3.400000       92.201000     -50.800000      0.634000   4963.600000  
25%       -1.800000       93.075000     -42.700000      1.344000   5099.100000  
50%        1.100000       93.749000     -41.800000      4.857000   5191.000000  
75%        1.400000       93.994000     -36.400000      4.961000   5228.100000  
max        1.400000       94.767000     -26.900000      5.045000   5228.100000

In [7]:
categorical = [var for var in df.columns if df[var].dtype=='O']
categorical

['job',
 'marital',
 'education',
 'default',
 'housing',
 'loan',
 'contact',
 'month',
 'day_of_week',
 'poutcome',
 'y']

In [8]:
df[categorical].isnull().sum()

job            0
marital        0
education      0
default        0
housing        0
loan           0
contact        0
month          0
day_of_week    0
poutcome       0
y              0
dtype: int64

In [9]:
numerical = [var for var in df.columns if df[var].dtype!='O']
numerical

['age',
 'duration',
 'campaign',
 'pdays',
 'previous',
 'emp.var.rate',
 'cons.price.idx',
 'cons.conf.idx',
 'euribor3m',
 'nr.employed']

In [10]:
df[numerical].isnull().sum()

age               0
duration          0
campaign          0
pdays             0
previous          0
emp.var.rate      0
cons.price.idx    0
cons.conf.idx     0
euribor3m         0
nr.employed       0
dtype: int64

In [11]:
def MultiLabelEncoder(columnlist,dataframe):
    for i in columnlist:
        labelencoder_X = LabelEncoder()
        dataframe[i] = labelencoder_X.fit_transform(dataframe[i])
columnlist = ['job','marital','education','default','housing','loan','contact','month','day_of_week','poutcome','y']
MultiLabelEncoder(columnlist,df)

In [12]:
df

age  job  marital  education  default  housing  loan  contact  month  \
0       56    3        1          0        0        0     0        1      6   
1       57    7        1          3        1        0     0        1      6   
2       37    7        1          3        0        2     0        1      6   
3       40    0        1          1        0        0     0        1      6   
4       56    7        1          3        0        0     2        1      6   
...    ...  ...      ...        ...      ...      ...   ...      ...    ...   
41183   73    5        1          5        0        2     0        0      7   
41184   46    1        1          5        0        0     0        0      7   
41185   56    5        1          6        0        2     0        0      7   
41186   44    9        1          5        0        0     0        0      7   
41187   74    5        1          5        0        2     0        0      7   

       day_of_week  duration  campaign  pdays  previous  poutcome  \
0                1       261         1    999         0         1   
1                1       149         1    999         0         1   
2                1       226         1    999         0         1   
3                1       151         1    999         0         1   
4                1       307         1    999         0         1   
...            ...       ...       ...    ...       ...       ...   
41183            0       334         1    999         0         1   
41184            0       383         1    999         0         1   
41185            0       189         2    999         0         1   
41186            0       442         1    999         0         1   
41187            0       239         3    999         1         0   

       emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  nr.employed  y  
0               1.1          93.994          -36.4      4.857       5191.0  0  
1               1.1          93.994          -36.4      4.857       5191.0  0  
2               1.1          93.994          -36.4      4.857       5191.0  0  
3               1.1          93.994          -36.4      4.857       5191.0  0  
4               1.1          93.994          -36.4      4.857       5191.0  0  
...             ...             ...            ...        ...          ... ..  
41183          -1.1          94.767          -50.8      1.028       4963.6  1  
41184          -1.1          94.767          -50.8      1.028       4963.6  0  
41185          -1.1          94.767          -50.8      1.028       4963.6  0  
41186          -1.1          94.767          -50.8      1.028       4963.6  1  
41187          -1.1          94.767          -50.8      1.028       4963.6  0  

[41188 rows x 21 columns]

In [13]:
X = df[categorical]
X = X.drop("y",axis=1)
X.columns

Index(['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact',
       'month', 'day_of_week', 'poutcome'],
      dtype='object')

In [14]:
y = df['y']
y

0        0
1        0
2        0
3        0
4        0
        ..
41183    1
41184    0
41185    0
41186    1
41187    0
Name: y, Length: 41188, dtype: int32

In [15]:
X_train,X_test, y_train,y_test = train_test_split(X,y,test_size=0.25,random_state=0)

In [16]:
X_train.shape

(30891, 10)

In [17]:
X_train.isnull().sum()

job            0
marital        0
education      0
default        0
housing        0
loan           0
contact        0
month          0
day_of_week    0
poutcome       0
dtype: int64

In [18]:
column_names = [X.columns]

In [19]:
categorical

['job',
 'marital',
 'education',
 'default',
 'housing',
 'loan',
 'contact',
 'month',
 'day_of_week',
 'poutcome',
 'y']

In [20]:
X

job  marital  education  default  housing  loan  contact  month  \
0        3        1          0        0        0     0        1      6   
1        7        1          3        1        0     0        1      6   
2        7        1          3        0        2     0        1      6   
3        0        1          1        0        0     0        1      6   
4        7        1          3        0        0     2        1      6   
...    ...      ...        ...      ...      ...   ...      ...    ...   
41183    5        1          5        0        2     0        0      7   
41184    1        1          5        0        0     0        0      7   
41185    5        1          6        0        2     0        0      7   
41186    9        1          5        0        0     0        0      7   
41187    5        1          5        0        2     0        0      7   

       day_of_week  poutcome  
0                1         1  
1                1         1  
2                1         1  
3                1         1  
4                1         1  
...            ...       ...  
41183            0         1  
41184            0         1  
41185            0         1  
41186            0         1  
41187            0         0  

[41188 rows x 10 columns]

In [21]:
# encoder = OneHotEncoder(column_names)
# X_train = encoder.fit_transform(X_train)
# X_test = encoder.transform(X_test)

In [22]:
# scaler = RobustScaler()

# X_train = scaler.fit_transform(X_train)

# X_test = scaler.transform(X_test)


# # X_train = pd.DataFrame(X_train, columns=X.columns)
# # X_test = pd.DataFrame(X_test, columns=X.columns)

In [26]:
gnb = GaussianNB()
gnb.fit(X_train, y_train)

GaussianNB(priors=None, var_smoothing=1e-09)

In [28]:
y_pred = gnb.predict(X_test)
y_pred

array([0, 0, 0, ..., 0, 0, 0])

In [30]:
print('Model accuracy score: {0:0.4f}'. format(accuracy_score(y_test, y_pred)))

Model accuracy score: 0.8374


In [31]:
print('Training-set accuracy score: {0:0.4f}'. format(accuracy_score(y_train, gnb.predict(X_train))))

Training-set accuracy score: 0.8422


In [32]:
print('Training set score: {:.4f}'.format(gnb.score(X_train, y_train)))

print('Test set score: {:.4f}'.format(gnb.score(X_test, y_test)))

Training set score: 0.8422
Test set score: 0.8374


In [33]:
y_test.value_counts()

0    9139
1    1158
Name: y, dtype: int64

In [35]:
print("Classification Report :")
print(classification_report(y_test, y_pred))

Classification Report :
              precision    recall  f1-score   support

           0       0.91      0.91      0.91      9139
           1       0.28      0.29      0.29      1158

    accuracy                           0.84     10297
   macro avg       0.60      0.60      0.60     10297
weighted avg       0.84      0.84      0.84     10297

